File: desktop/aes_pack.py
import os
import pyaes
import struct
import numpy as np


class AesPack:
    def __init__(self, key):
        self.bin_nonce = os.urandom(8)
        self.nonce, = struct.unpack("!Q", self.bin_nonce)
        self.crypt_seq = bytearray(0)
        self.aes = pyaes.AES(key)

        self.decrypt_nonce = None
        self.decrypt_seq = bytearray(0)

    def crypt(self, bts):
        check_sum = self.checksum(bts)
        bts += struct.pack("!L", check_sum)

        self.crypt_seq = self.expand_crypt_seq(self.crypt_seq, len(bts), self.nonce)
        return self.bin_nonce + self.do_xor(bts, self.crypt_seq)

    def decrypt(self, bts):
        if len(bts) < 9:
            return None

        nonce, = struct.unpack("!Q", bts[:8])
        if self.decrypt_nonce != nonce:
            self.decrypt_nonce = nonce
            self.decrypt_seq = bytearray(0)

        self.decrypt_seq = self.expand_crypt_seq(self.decrypt_seq, len(bts)-8, nonce)
        decrypted = self.do_xor(bts[8:], self.decrypt_seq)

        stored_check_sum, = struct.unpack("!L", decrypted[-4:])
        res = decrypted[:-4]
        check_sum = self.checksum(res)
        if check_sum != stored_check_sum:
            return None

        return res

    def expand_crypt_seq(self, crypt_seq, seq_len, nonce):
        if seq_len <= len(crypt_seq):
            return crypt_seq

        chunks_count = int((seq_len+15)/16)
        existent_chunks_count = int(len(crypt_seq)/16)

        crypt_seq += bytearray((chunks_count - existent_chunks_count)*16)

        for chunk in range(existent_chunks_count, chunks_count):
            c = struct.pack("!Q", nonce + chunk) + struct.pack("!Q", nonce + chunk + 1)
            crypt_seq[chunk*16:(chunk+1)*16] = self.aes.encrypt(c)

        #print(f'expand {chunks_count=} {existent_chunks_count=} {nonce=}')

        return crypt_seq

    @staticmethod
    def do_xor(bts, crypt_seq):
        np_bts = np.frombuffer(bts, dtype=np.uint8)
        crypt_bts = np.frombuffer(crypt_seq[:len(bts)], dtype=np.uint8)
        res = np_bts ^ crypt_bts
        #print(f'{np_bts=} {crypt_bts=} {res=}')
        return res.tobytes()

    @staticmethod
    def checksum(bts):
        np_bts = np.frombuffer(bts, dtype=np.uint8)
        return int(np.sum(np_bts) % 0x100000000)

File: android/data/src/main/java/com/robocart/data/network/AesPack.kt
package com.robocart.data.network

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec

class AesPack(key: ByteArray) {
    private val encryptNonceBytes = ByteArray(8).also { SecureRandom().nextBytes(it) }
    private val encryptNonce = readUInt64(encryptNonceBytes, 0)
    private var encryptSeq = ByteArray(0)
    private var decryptNonce: ULong? = null
    private var decryptSeq = ByteArray(0)
    private val aesCipher = Cipher.getInstance("AES/ECB/NoPadding").apply {
        init(Cipher.ENCRYPT_MODE, SecretKeySpec(key.copyOf(), "AES"))
    }

    fun crypt(bytes: ByteArray): ByteArray {
        val payload = ByteArray(bytes.size + 4)
        System.arraycopy(bytes, 0, payload, 0, bytes.size)
        writeUInt32(payload, bytes.size, checksum(bytes))
        encryptSeq = expandCryptSeq(encryptSeq, payload.size, encryptNonce)
        val encryptedPayload = doXor(payload, encryptSeq)
        return encryptNonceBytes + encryptedPayload
    }

    fun decrypt(bytes: ByteArray): ByteArray? {
        if (bytes.size < 9) {
            return null
        }

        val nonce = readUInt64(bytes, 0)
        if (decryptNonce != nonce) {
            decryptNonce = nonce
            decryptSeq = ByteArray(0)
        }

        val encrypted = bytes.copyOfRange(8, bytes.size)
        decryptSeq = expandCryptSeq(decryptSeq, encrypted.size, nonce)
        val decrypted = doXor(encrypted, decryptSeq)
        if (decrypted.size < 4) {
            return null
        }

        val storedChecksum = readUInt32(decrypted, decrypted.size - 4)
        val plain = decrypted.copyOfRange(0, decrypted.size - 4)
        if (checksum(plain) != storedChecksum) {
            return null
        }

        return plain
    }

    private fun expandCryptSeq(currentSeq: ByteArray, requiredLen: Int, nonce: ULong): ByteArray {
        if (requiredLen <= currentSeq.size) {
            return currentSeq
        }

        val chunksCount = (requiredLen + 15) / 16
        val existingChunks = currentSeq.size / 16
        val seq = currentSeq.copyOf(chunksCount * 16)
        for (chunk in existingChunks until chunksCount) {
            val first = nonce + chunk.toULong()
            val second = nonce + (chunk + 1).toULong()
            val block = ByteBuffer.allocate(16)
                .order(ByteOrder.BIG_ENDIAN)
                .putLong(first.toLong())
                .putLong(second.toLong())
                .array()
            val encryptedBlock = aesCipher.doFinal(block)
            System.arraycopy(encryptedBlock, 0, seq, chunk * 16, 16)
        }
        return seq
    }

    companion object {
        fun checksum(bytes: ByteArray): UInt {
            var sum = 0UL
            for (value in bytes) {
                sum += value.toUByte().toULong()
            }
            return (sum and 0xFFFF_FFFFu).toUInt()
        }

        fun doXor(bytes: ByteArray, cryptSeq: ByteArray): ByteArray {
            val result = ByteArray(bytes.size)
            for (i in bytes.indices) {
                result[i] = (bytes[i].toInt() xor cryptSeq[i].toInt()).toByte()
            }
            return result
        }

        private fun writeUInt32(target: ByteArray, offset: Int, value: UInt) {
            target[offset] = ((value.toLong() ushr 24) and 0xFF).toByte()
            target[offset + 1] = ((value.toLong() ushr 16) and 0xFF).toByte()
            target[offset + 2] = ((value.toLong() ushr 8) and 0xFF).toByte()
            target[offset + 3] = (value.toLong() and 0xFF).toByte()
        }

        private fun readUInt32(source: ByteArray, offset: Int): UInt {
            val a = source[offset].toUByte().toUInt() shl 24
            val b = source[offset + 1].toUByte().toUInt() shl 16
            val c = source[offset + 2].toUByte().toUInt() shl 8
            val d = source[offset + 3].toUByte().toUInt()
            return a or b or c or d
        }

        private fun readUInt64(source: ByteArray, offset: Int): ULong {
            var value = 0UL
            for (i in 0 until 8) {
                value = (value shl 8) or source[offset + i].toUByte().toULong()
            }
            return value
        }
    }
}
